acra/security-controls/masking/_index.md

```diff
@@ -2,3 +2,56 @@
 title: Masking
 bookCollapseSection: true
 ---
+
+# Masking
+
+Partial column encryption.
+Hides encrypted part from those who don't have right decryption keys.
+In case of decryption failure, encrypted data will be replaced with a placeholder.
+
+## AcraServer configuration
+
+In configuration file, passed by `--encryptor_config_file` flag, you can individually configure
+masking for any field of any table, if that field can be encrypted.
+These options are accepted:
+
+<!-- Config struct lives in encryptor/config/encryptionSettings.go -->
+* `masking` — text string that will be shown in unsuccessful decryption responses instead of encrypted data
+* `plaintext_length` — how many bytes should be left unencrypted
+* `plaintext_side` — which side of column should contain unencrypted part of value
+  * `left`
+  * `right`
+* `crypto_envelope` — which crypto container to use
+  * [`acrastruct`]({{< ref "acra/acra-in-depth/data-structures/_index.md#acrastruct" >}}) (default)
+  * [`acrablock`]({{< ref "acra/acra-in-depth/data-structures/_index.md#acrablock" >}})
+* `reencrypting_to_acrablocks` — whether to automatically re-encrypt values stored
+  in [AcraStructs]({{< ref "acra/acra-in-depth/data-structures/_index.md#acrastruct" >}})
+  to [AcraBlocks]({{< ref "acra/acra-in-depth/data-structures/_index.md#acrablock" >}})
+<!-- TODO add link to page where colemn encryption settings are described in general, with client_id, zone_id etc -->
+
+<!-- Config validation func lives in masking/common/patterns.go -->
+In order to enable masking you will need:
+* non-empty value for `masking`
+* `plaintext_length` ⩾ 0
+* valid value for `plaintext_side` (`left` or `right`)
```

Comment:
I find the “commented example configuration” to be one of the best ways to document configuration files.

For example:

```
{
    // (required, must be non-empty)
    // Replacement string to use for masking out encrypted data.
    // If decryption fails, this string will be show instead of the masked portion of the data.
    "masking": "XXXXXXXXXXXX",

    // (required, must be non-negative)
    // How many bytes of plaintext to leave unencrypted.
    "plaintext_length": 4,

    // (required, allowed values: "left", "right")
    // Which side of the plaintext to leave unencrypted.
    // "plaintext_length" bytes will be retained from left or right side of the data cell.
    "plaintext_side": "right",

    // (optional, default: "acrastruct", allowed values: "acrastruct", "acrablock")
    // Which cryptographic container to use for data encryption.
    "crypto_envelope": "acrastruct",

    // (optional, default: false)
    // If true, data stored in AcraStructs will be transparently reencrypted into AcraBlocks.
    "reencrypting_to_acrablocks": false,
}
```

This also shows one of the reasons why JSON is not particularly suited for configuration files, since the spec does not include comments. Sure, *your* parser might support and ignore them but others might not. This creates unnecessary friction, for example, when you want to add syntax highlighting to your documentation but that parser does not want to recognize comments.

Reply:
Took your variant. But had to split it a bit because of horizontal scroll we don't want. And made it YAML because we actually use this format for configuration.

> the spec does not include comments

That's why someone invented JSON5 :)

Reply:
> made it YAML because we actually use this format for configuration

I guess I got bamboozled by the provided examples:

```
{ masking: "*", plaintext_length: 5, plaintext_side: "left" }
```

which certainly looks JSON-like to me.

Meanwhile, I don't think that

```yaml
{
  masking: "???",
  plaintext_length: 0,
  plaintext_side: "left"
}
```

is valid YAML.

Shouldn't it be something like this instead?

```yaml
masking: "???"
plaintext_length: 0
plaintext_side: "left"
```

Reply:
Every JSON is valid YAML AFAIK. This one seems valid too, checked on https://yamlvalidator.com/.
But your words make sense, guess it's better to remove `{` `}` OR make it look like a part of column encryption settings:
```yaml
# in config description

schemas:
  - table: foo
    encrypted:
      - column: first_name
        # description
        masking: "XXXX"
        # description
        plaintext_length: 1
        # description
        plaintext_side: "left"
        # same with crypto_envelope and reencrypting_to_acrablocks

# in examples

masking: "*"
plaintext_length: 1
plaintext_side: "left"
```